File: sudoq-app/sudoqmodel/src/main/java/de/sudoq/model/solverGenerator/solver/helper/SolveHelper.java
package de.sudoq.model.solverGenerator.solver.helper;

import de.sudoq.model.solverGenerator.solution.SolveDerivation;
import de.sudoq.model.solverGenerator.solver.SolverSudoku;
import de.sudoq.model.solvingAssistant.HintTypes;

/**
 * Dieses Interface definiert eine einheitliche Schnittstelle für verschiedene Vorgehensweisen zum Lösen eines Sudokus.
 * Alle Helper aktualisieren bei erfolgreicher Anwendung der Vorgehensweise die Kandidatenlisten der Felder und
 * schränken so die Möglichkeiten ein Symbol in ein Feld einzutragen ein, womit i. A. andere Helper ihre Vorgehensweise
 * anwenden können.
 */
public abstract class SolveHelper {
	/** Attributes */

	/**
	 * Das Sudoku, auf welches dieser SolveHelper seine Vorgehensweise anwendet
	 */
	protected SolverSudoku sudoku;

	/**
	 * Difficulty score of this helper e.g. 1 for `naked single` and 100 for `xwing` added up they give a score for the difficulty of a sudoku
	 */
	private int complexityScore;

	/**
	 * Die Herleitung des letzten update-Schrittes;
	 */
	protected SolveDerivation lastDerivation;

	public HintTypes hintType; //public only for debugging


	/** Constructors */

	/**
	 * Creates a new SolveHelper for the specified Sudoku with the specified complexity.
	 * 
	 * @param sudoku
	 *            sudoku to find a helper for. Mustn't be null
	 * @param complexity
	 *            desired complexity for the final sudoku. Must be {@code >= 0}.
	 */
	protected SolveHelper(SolverSudoku sudoku, int complexity) {
		assert sudoku != null;
		assert complexity >= 0 : "complexity < 0 : " + complexity;

		this.sudoku = sudoku;
		this.complexityScore = complexity;
	}

	/** Methods */

	/**
	 * Versucht die entsprechende Vorgehensweise solange auf das im konkreten SolveHelper gespeicherte Sudoku
	 * anzuwenden, bis sie zum ersten Mal zum Erfolg führt oder sie nirgendwo anwendbar ist. Kann die Vorgehensweise
	 * erfolgreich angewandt werden und wurde true für den Parameter buildDerivation übergeben, so wird dieses Vorgehen
	 * als SolveDerivation gespeichert und kann über getDerivation abgerfragt werden. Führt die Vorgehensweise bei dem
	 * spezifizierten Sudoku nicht zum Erfolg, so wird false zurückgegeben.
	 * 
	 * @param buildDerivation
	 *            Bestimmt, ob eine Herleitung erstellt werden soll, welche durch die getDerivation-Methode abgerufen
	 *            werden kann oder nicht
	 * @return true, falls dieser Helper angewendet werden konnte, false falls nicht
	 */
	abstract public boolean update(boolean buildDerivation);

	/**
	 * Gibt die Herleitung des letzten update-Schrittes zurück, sofern dieser mit dem Parameter buildDerivation
	 * aufgerufen wurde und erfolgreich war. Ansonsten wird false zurückgegeben.
	 * 
	 * @return die Herleitung des letzten update-Schritte, falls vorhanden, sonst null
	 */
	public SolveDerivation getDerivation() {
		return this.lastDerivation;
	}

	/**
	 * Gibt die Schwierigkeit der Anwendbarkeit dieses Helfers zurück. Dieser ist mit dem Konstruktor zu setzen.
	 * 
	 * @return difficulty score of this helper
	 *
	 */
	public int getComplexityScore() {
		return this.complexityScore;
	}

	public HintTypes getHintType(){return hintType;}
}

File: sudoq-app/sudoqmodel/src/main/java/de/sudoq/model/solverGenerator/solver/helper/SolveHelper.kt
package de.sudoq.model.solverGenerator.solver.helper

import de.sudoq.model.solverGenerator.solution.SolveDerivation
import de.sudoq.model.solverGenerator.solver.SolverSudoku
import de.sudoq.model.solvingAssistant.HintTypes

/**
 * A Solution Technique for a Sudoku.
 * If application is successful the candidates of the cells are reduced with can help other helpers.
 *
 * @property sudoku SolverSudoku on which SolveHelper operates
 * @param complexity desired complexity for the final sudoku. Must be `>= 0`.
 */
abstract class SolveHelper protected constructor(
    @JvmField protected var sudoku: SolverSudoku,
    complexity: Int
) {

    /**
     * Difficulty score of this helper e.g.
     *  1 for `naked single` and
     *  100 for `xwing` added up they give a score for the difficulty of a sudoku
     */
    val complexityScore: Int

    init {
        require(complexity >= 0) { "complexity < 0 : $complexity" }
        complexityScore = complexity
    }

    /**
     * The derivation of the last update step if
     * it was called with buildDerivation=true and if it was successful.
     *
     * @return Derivation of the last update step, ofr null if not available
     */
    var derivation: SolveDerivation? = null
        protected set

    var hintType: HintTypes? = null //public only for debugging

    /**
     * Tries to apply this helper to the Sudoku until it succeeds for the first time or cannot be applied.
     * If the helper is successful and was called with `buildDerivation == true` the a derivation is saved.
     * It can be queried with getDerivation.
     *
     * @param buildDerivation Specifies if a derivation should be stored.
     * @return true, falls the helper could be applied(i.e. change the sudoku), false if not
     */
    abstract fun update(buildDerivation: Boolean): Boolean
}